Add IPv6 BIG TCP support

NikAleksandrov opened the pull request:
Add a new IPv6 BIG TCP[1] option and infrastructure which allows to change a device's GRO and GSO max sizes depending on their current values and the new option (enable-ipv6-big-tcp). BIG TCP is used to allow the network stack to cook bigger TSO/GRO packets (512k is the new maximum, 64k was the old), it can be used only with IPv6 and improves latency and performance due to the reduced number of packets that traverse the stack. It works by adding a new temporary Hop by Hop header right after the IPv6 header, it is stripped before the packet is sent on the wire.

The defaults (65536) are set only if necessary (i.e. if the GSO/GRO max sizes were previously changed), when the option is enabled they get set to 196608. The bpf_dynptr_data helper is used to probe for 5.19+ kernel. BIG TCP can be used only in BPF host routing mode, without encryption and tunneling, there are checks that enforce compatibility. If BIG TCP cannot tune all external interfaces then it will revert back to the defaults. Currently it is supported only on mlx4/mlx5 and veth devices, also any device that inherits gro/gso maximum sizes from others (e.g. bonding should work).

LWN also has a nice BIG TCP summary available [here](https://lwn.net/Articles/884104/)

If the new BIG TCP Cilium option is enabled and initialized it can be seen in the logs:
level=info msg="Setting up IPv6 BIG TCP" subsys=big-tcp
level=info msg="Setting gso_max_size to 196608 and gro_max_size to 196608" device=eth0 subsys=big-tcp

Netperf benchmarks:

```
- TCP_RR
MIN_LATENCY  P90_LATENCY  P99_LATENCY  THROUGHPUT (trans/sec)

BIG TCP Disabled
60            91           136         13087.53

BIG-TCP Enabled
38            76           109         15629.73

- TCP_STREAM (Mbps)
16386.21 (BIG TCP disabled)
23181.41 (BIG TCP enabled)
```

Standalone tests which use a Kind cluster are added. BIG TCP currently only works with veth and mellanox devices
so we need a Kind cluster to test it properly, I've left a TODO note that it should be migrated to the new infra ones it
lands.

Note that pods need to  be restarted when the option is changed due to the veth configuration which gets applied when the
devices are created and before they're put in the target netns.

Patch-set overview:
 Patch 01 - adds basic BIG TCP infrastructure allowing to enable/disable it and check for requirements
 Patch 02 - adds GRO/GSO max sizes to the daemon configuration status so they can be exposed for veth device configuration
 Patch 03 - adds the new "--enable-ipv6-big-tcp" option to the daemon
 Patch 04 - uses the exposed GRO/GSO max sizes to configure the newly created pod veth devices
 Patch 05 - adds helm support for the new option (enableIPv6BIGTCP)
 Patch 06 - exposes the IPv6 BIG TCP state in "cilium status" (IPv6 BIG TCP:            Enabled/Disabled)
 Patch 07 - documents the new feature in the tuning docs
 Patch 08 - adds BIG TCP tests which provision a Kind cluster, verify the option is properly enabled and run netperf

Requirements to enable:
* Kernel >= 5.19
* eBPF Host-Routing
* eBPF-based kube-proxy replacement
* eBPF-based masquerading
* Tunneling and encryption disabled
* Supported NICs: mlx4, mlx5

Thanks,
  Nik

[1] https://lore.kernel.org/netdev/20220513183408.686447-3-eric.dumazet@gmail.com/T/

Commit:
vendor: update vishvananda/netlink/

Update vishvananda/netlink to support setting IFLA_GRO_MAX_SIZE and
IFLA_GSO_MAX_SIZE which is needed for BIG TCP.

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

Commit:
datapath: connector: veth: allow GRO/GSO max size configuration

Allow configuring GRO/GSO max sizes when setting up veth devices. These
are needed to enable BIG TCP support. They are configured only if > 0.
Pass the configured values when setting up veths.

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

Commit:
datapath: linux: add IPv6 BIG TCP support

Add basic IPv6 BIG TCP infrastructure which allows to change a device's
GRO and GSO max sizes depending on their current value and a new
enable-ipv6-big-tcp option. BIG TCP is used to allow the network stack
to cook bigger TSO/GRO packets (512k is the new maximum, 64k was the
old), it can be used only with IPv6 and improves latency and performance
due to the reduced number of packets that traverse the stack. It works
by adding a new temporary Hop by Hop header right after the IPv6 header.
Note that may break some eBPF programs which assume TCP header follows
immediately the IPv6 header.

The defaults (65536) are set only if necessary (i.e. if the GSO/GRO max
sizes were previously changed), when the option is enabled they get set
to 196608. The bpf_dynptr_data helper is used to probe for 5.19+ kernel.
BIG TCP can be used only in BPF host routing mode, without encryption and
tunneling, there are checks that enforce compatibility. If BIG TCP cannot
tune all external interfaces then it will revert back to the defaults.
Currently it is supported only on mlx4/mlx5 and veth devices, also any
device that inherits gro/gso maximum sizes from others (e.g. bonding
should work).

If BIG TCP is enabled and initialized it can be seen in the logs:
level=info msg="Setting up IPv6 BIG TCP" subsys=big-tcp
level=info msg="Setting gso_max_size to 196608 and gro_max_size to 196608" device=eth0 subsys=big-tcp

Benchmarks:

- TCP_RR
MIN_LATENCY  P90_LATENCY  P99_LATENCY  THROUGHPUT (trans/sec)

BIG TCP Disabled
60           91           136          13087.53

BIG-TCP Enabled
38           76           109          15629.73

- TCP_STREAM
16386.21 (BIG TCP disabled)
23181.41 (BIG TCP enabled)

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

Commit:
openapi: add GRO/GSO max size to the daemon configuration status

Add new GRO/GSO max size fields to the daemon configuration status API.
We need to expose them so they can be configured on pod veth devices.

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

Commit:
daemon: add IPv6 BIG TCP option and config support

Add daemon support for the new option, initialize it in its NewDaemon
call and expose the initialized GRO/GSO max sizes (BIG TCP config) through
the daemon configuration status.

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

Commit:
cilium, helm: add IPv6 BIG TCP

Add Helm setting for IPv6 BIG TCP (enableIPv6BIGTCP) which defaults to
false. Used "make -C install/kubernetes cilium/values.yaml" to
autogenerate the values.yaml file.

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

Commit:
test: add BIG TCP tests

Add standalone BIG TCP tests. They use a Kind cluster and setup Cilium
with BIG TCP enabled. Then verify that gso_max_size is set properly (not
verifying gro_max_size due to availability in iproute2, gso_max_size has
been supported for a long time and if it was set properly that means gro
max size was also set). And lastly perform a netperf TCP_RR test between
the client and server netperf pods. The test needs veth devices for BIG
TCP support and Kind is the natural choice, when we add e2e Kind testing
infra these tests can be moved and integrated (left a TODO note in the
test as well).

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

Commit:
cilium: expose BIG TCP state in status output

Add current BIG TCP state to the Status Response model and expose it
in "cilium status". The struct naming (IPV6BigTCP) is due to the
automatic generation.

Output looks like:
$ kubectl -n kube-system exec cilium-rmxzw -- cilium status
...
IPv6 BIG TCP:            Enabled
...

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

Commit:
docs: tuning: add a paragraph about BIG TCP

Add a new entry explaining the BIG TCP feature, its requirements, how to
enable it and how to validate if it was successfully enabled.

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

borkmann commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -206,7 +206,6 @@ azure:
   # Note that this is incompatible with AKS clusters created in BYOCNI mode: use
   # AKS BYOCNI integration (`aksbyocni.enabled`) instead.
   enabled: false
-  # usePrimaryAddress: false
```
unrelated?

NikAleksandrov replied:
it was because of "make -C install/kubernetes cilium/values.yaml"
one of the tests was failing without it

NikAleksandrov replied:
anyway, it seems after rebasing on master that isn't needed anymore so I've dropped it

borkmann reviewed (approved):
LGTM! Great work 🚀

brb commented on test/bigtcp/test.sh:
```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
```
You might want to run this test as part of the CI L4LB. To make it happen, you need to add exec the test.sh here: https://github.com/cilium/cilium/blob/master/.github/workflows/tests-l4lb.yaml#L198

NikAleksandrov replied:
ah nice catch, I'll update the test commit to include it, thanks!

brb reviewed (changes requested):
Nice PR! Reviewed the testing bits.

jrfastab commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -0,0 +1,152 @@
+		(!option.Config.EnableIPv6BIGTCP &&
+			int(attrs.GROMaxSize) <= GROMaxSize &&
+			int(attrs.GSOMaxSize) <= GSOMaxSize) {
+		return nil
```
The second case here, 
```
!EnableIPv6BIGTCP && attrs.GROMaxSize <= GROMaxSize && GSOMaxSize <= GSOMaxSize
```
this is an error right? The user tried to set GRO/GSO larger than the max should we return an error?

NikAleksandrov replied:
no, that's to avoid changing gso/gro when not necessary, we have the default case where big tcp is disabled and the case where it was enabled and then disabled, this is used to match the first one

NikAleksandrov replied:
I'll add a comment

jrfastab commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -0,0 +1,152 @@
+		return err
+	}
+
+	return netlink.LinkSetGSOMaxSize(link, GSOMaxSize)
```
If this errors should we unwinde the GROMaxSize case so this routine doesn't leave us in some partial state.

jrfastab replied:
ok I see you roll back in the caller now. I guess this is OK. Maybe add a comment to the function so next dev that comes along knows they need to do rollback from caller as well.

NikAleksandrov replied:
sure, will do

NikAleksandrov replied:
added a comment to make sure devs know about rolling back any changes that have been done

jrfastab commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -0,0 +1,152 @@
+	}
+	if !kernelGood {
+		if option.Config.EnableIPv6BIGTCP {
+			log.Fatalf("Cannot enable --%s, needs kernel 5.19 or newer",
```
Do we really want to fatalf here? This seems extreme to me. In alpo for example we have a mix of kernel versions and we might upgrade some nodes to 5.19 when its available but this will stop us from enabling it in configMap. I would suggest if we really want to fatalf here we should have 'strict' and 'probe' modes.

NikAleksandrov replied:
we can just disable the option with a log msg, but people must look for it to verify what happened, I think a hard error is better so they can't miss it, that's why it fatalfs in the requirement checks

jrfastab replied:
it will break deployment into our dogfooding cluster though.

NikAleksandrov replied:
ok, I've pushed a change to make it a warning with a message why it couldn't be enabled if the kernel is incompatible

jrfastab commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -0,0 +1,152 @@
+		}
+		if option.Config.EnableHostLegacyRouting {
+			log.Fatal("IPv6 BIG TCP is not supported with legacy host routing")
+		}
```
OTOH these fatal events seem OK to me because this is a config error that needs to be resolved.

NikAleksandrov replied:
ah right, that's what I meant

jrfastab commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -0,0 +1,152 @@
+
+	if err != nil {
+		bigtcpConfig.groMaxSize = defaultGROMaxSize
+		bigtcpConfig.gsoMaxSize = defaultGSOMaxSize
```
Looks like default*Size is just a var  const. Probably to be save should read these out at top of function so if something else has configured values we don't whack them here on error. GKE for example does some /proc settings I would guess others do as well.

NikAleksandrov replied:
the problem is once it's enabled we don't know what it was before that, we could record them but it seems like an overkill, gso and gro max sizes are rarely changed from their 64k defaults,
I'll think about it more and will check if any of the vendors changes them

edit: ahh I missed the error part, yeah we could do that for the error path. TBH I very much doubt anyone is reducing gro or gso max sizes. :)

NikAleksandrov replied:
By the way note that these settings are per-device and can only be changed via netlink calls (they aren't exported to /proc or /sys)

NikAleksandrov replied:
I left a comment in big tcp's docs about the assumptions we make and the logic of enable/disable (when Cilium actually does any value changes). I think that should be ok for the time being, if we hit any issues we can revisit the logic and perhaps persist the old values to be sure that we need to change them to anything different.

jrfastab commented on daemon/cmd/daemon_main.go:
```diff
@@ -725,6 +725,9 @@ func initializeFlags() {
 	flags.Bool(option.EnableIPMasqAgent, false, "Enable BPF ip-masq-agent")
 	option.BindEnv(option.EnableIPMasqAgent)
 
+	flags.Bool(option.EnableIPv6BIGTCP, false, "Enable IPv6 BIG TCP option which increases device's maximum GRO/GSO limits")
```
maybe make a tri-state to handle case where we fatal fail if we can't configure this?

NikAleksandrov replied:
it already does that in most cases when requirements can't be met, am I missing some case or did I misunderstand you?

brb reviewed (approved):
The CI changes LGTM. Thanks.

gandro commented on cilium-health/launch/endpoint.go:
```diff
@@ -272,7 +272,7 @@ func LaunchAsEndpoint(baseCtx context.Context,
 	switch option.Config.DatapathMode {
 	case datapathOption.DatapathModeVeth:
-		_, epLink, err := connector.SetupVethWithNames(vethName, epIfaceName, mtuConfig.GetDeviceMTU(), info)
+		_, epLink, err := connector.SetupVethWithNames(vethName, epIfaceName, mtuConfig.GetDeviceMTU(), 0, 0, info)
```
Any particular reason as to why you are not passing in `d.bigtcpConfig` here? `LaunchAsEndpoint` is called from the Daemon, so it should be fairly straight forward to get it.

Now I understand that the health does not send big packets, so there is absolutely no performance benefit to do it. But on the other hand, the health endpoint should imitate the regular pod setup as much as possible, so we catch connectivity issues there early. I'm fine to keep the code as is, but I am curious if you have given this any thought.

NikAleksandrov replied:
IIRC I didn't want to add any additional unnecessary direct dependencies on datapath/linux/bigtcp and to lesser extent also to be consistent with the rest of the code which takes two integers for GSO and GRO.

gandro replied:
Not sure I follow. You can decouple the package dependency by adding an interface for `Get{GRO,GSO}MaxSize`, similar to e.g. the `routingConfigurer` interface. Or you could just pass in two integers if you find that more consistent?

NikAleksandrov replied:
I don't think the health monitor needs large gso/gro limits as you've also noted, keeping it as-is seems marginally simpler, but I don't have a strong preference and having config parity sounds good so I  don't mind passing the real values. I'll push the change after running a few tests locally.

NikAleksandrov replied:
I've pushed the changes in this commit which also adds the support for configuring gso/gro limits on a veth pair.
Thanks for the review!

gandro replied:
Awesome, thanks Nikolay!

gandro reviewed (changes requested):
Reviewed Helm and Health. Both look good to me, one high-level question around health however.

(edit: I misclicked. I meant to approve, no blocking changes requested)

qmonnet commented on Documentation/operations/performance/tuning.rst:
```diff
@@ -37,6 +37,55 @@ in any of the Cilium pods and look for the line reporting the status for
+IPv6 BIG TCP
+============
+
+IPv6 BIG TCP allows the network stack to prepare larger TSO (transmit) and GRO
```
Nit: s/TSO/GSO/ ? I know “TSO” exists, but the rest of the document only mentions GSO (and conversely, it uses GSO later without mentioning it's for transmitting).

NikAleksandrov replied:
That was intentional actually because GSO segments are later passed down for TSO, but I agree it's best to be consistent.
I'll change it, thanks!

qmonnet commented on Documentation/operations/performance/tuning.rst:
```diff
@@ -37,6 +37,55 @@ in any of the Cilium pods and look for the line reporting the status for
+
+**Requirements:**
+
+* Kernel >= 5.19
```
Maybe worth adding a line on kernel requirements in https://docs.cilium.io/en/latest/operations/system_requirements/#required-kernel-versions-for-advanced-features as well?

NikAleksandrov replied:
sounds good, I'll add that as well

qmonnet commented on test/bigtcp/test.sh:
```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
```
Please add the license/copyright header. I know not all bash scripts have it, but let's try to add it for new files.

NikAleksandrov replied:
ack, will do

qmonnet reviewed (approved):
The code looks good. Nice work, thanks! PR is well split and easy to review.

I've got some nitpicks on the doc (non-blocking), please see below.

Commit:
docs: system requirements: add BIG TCP kernel requirements

Add BIG TCP's kernel requirements to "Required Kernel Versions for
Advanced Features"

Signed-off-by: Nikolay Aleksandrov <nikolay@isovalent.com>

tommyp1ckles commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -107,6 +107,7 @@ cilium-agent [flags]
       --enable-ipv4-fragment-tracking                           Enable IPv4 fragments tracking for L4-based lookups (default true)
       --enable-ipv4-masquerade                                  Masquerade IPv4 traffic from endpoints leaving the host (default true)
       --enable-ipv6                                             Enable IPv6 support (default true)
+      --enable-ipv6-big-tcp                                     Enable IPv6 BIG TCP option which increases device's maximum GRO/GSO limits
```
```suggestion
      --enable-ipv6-big-tcp                                     Enable IPv6 BIG TCP option which increases device's maximum GRO/GSO limits (default false)
```

NikAleksandrov replied:
ack, I've added the default

NikAleksandrov replied:
actually no, the default was there in the beginning and when building the documentation checks remove it, if I add it back the docs test test below fails so I've removed it again

tommyp1ckles commented on pkg/datapath/connector/veth.go:
```diff
@@ -123,6 +123,28 @@ func SetupVethWithNames(lxcIfName, tmpIfName string, mtu int, ep *models.Endpoin
 
+	if groMaxSize > 0 {
+		if err = netlink.LinkSetGROMaxSize(hostVeth, groMaxSize); err != nil {
+			return nil, nil, fmt.Errorf("unable to set GRO max size to %q: %s",
```
I would use error wrapping for these, same goes for the 3 returns with fmt.Errorf below.
```suggestion
			return nil, nil, fmt.Errorf("unable to set GRO max size to %q: %w",
```

NikAleksandrov replied:
ack, changed

tommyp1ckles commented on pkg/datapath/connector/veth.go:
```diff
@@ -123,6 +123,28 @@ func SetupVethWithNames(lxcIfName, tmpIfName string, mtu int, ep *models.Endpoin
+				lxcIfName, err)
+		}
+		if err = netlink.LinkSetGROMaxSize(peer, groMaxSize); err != nil {
+			return nil, nil, fmt.Errorf("unable to set GRO max size to %q: %s",
```
Should we distinguish between the host & peer link in either of these error cases messages?

NikAleksandrov replied:
No, we don't care which one failed. This is handled the same for both by the driver, if one fails then the other will fail as well.

tommyp1ckles commented on pkg/datapath/connector/veth.go:
```diff
@@ -123,6 +123,28 @@ func SetupVethWithNames(lxcIfName, tmpIfName string, mtu int, ep *models.Endpoin
+				lxcIfName, err)
+		}
+		if err = netlink.LinkSetGSOMaxSize(peer, gsoMaxSize); err != nil {
+			return nil, nil, fmt.Errorf("unable to set GSO max size to %q: %s",
```
Same question as above about distinguishing between links.

tklauser commented on api/v1/openapi.yaml:
```diff
@@ -2821,6 +2830,16 @@ definitions:
       forwarded:
         description: Number of messages forwarded
         type: integer
+  IPV6BigTCP:
```
Nit:

```suggestion
  IPv6BigTCP:
```

NikAleksandrov replied:
this was changed automatically by something that checks the openapi file, I had it as IPv6BigTCP and it failed a test here

tklauser commented on daemon/cmd/daemon.go:
```diff
@@ -195,6 +196,9 @@ type Daemon struct {
 	controllers *controller.Manager
+
+	// BIG-TCP config values
+	bigtcpConfig bigtcp.Configuration
```
```suggestion
	bigTCPConfig bigtcp.Configuration
```

NikAleksandrov replied:
sure, why not :)

tklauser commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -0,0 +1,158 @@
+	defaultGSOMaxSize int = 65536
+
+	BIGTCPGROMaxSize int = 196608
+	BIGTCPGSOMaxSize int = 196608
```
These do not seem to be used outside the package, so:

```suggestion
	bigTCPGROMaxSize int = 196608
	bigTCPGSOMaxSize int = 196608
```

NikAleksandrov replied:
They weren't big for export purposes (although that happens), but because that's the real name of the feature BIG-TCP.
Anyway, I'll make them small to avoid exporting them

tklauser commented on pkg/datapath/linux/bigtcp/bigtcp.go:
```diff
@@ -0,0 +1,158 @@
+	defaultGSOMaxSize int = 65536
+
+	BIGTCPGROMaxSize int = 196608
+	BIGTCPGSOMaxSize int = 196608
```
It's more common in Go to declare numeric constants as untyped. That way they can be used with arbitrary integer types depending on context. Any particular reason these need to be `int`s? I think making them untyped would allow to get rid of the type conversions in `setGROGSOMaxSize` below.

```suggestion
	defaultGROMaxSize = 65536
	defaultGSOMaxSize = 65536

	BIGTCPGROMaxSize = 196608
	BIGTCPGSOMaxSize = 196608
```

NikAleksandrov replied:
no particular reason or preference, I'll change them

NikAleksandrov replied:
Actually they aren't used in setGROGSOMaxSize() directly, maxGRO/GSO are passed as integer arguments
because the netlink library takes GSO/GRO values as ints.

tklauser replied:
Right, the type conversions in setGROGSOMaxSize are still needed then. But I think the consts can remain untyped.

NikAleksandrov replied:
Yep, they can. I'll push the changes in a minute

rolinh reviewed (approved):
LGTM for the vendor and API changes :+1:
